File: python_app/models/position.py
"""Position Record model - translated from POSREC.cpy copybook.

Mirrors the COBOL POSITION-RECORD structure with composite key
(portfolio_id, date, investment_id) and position data fields.
"""

from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, Field


class PositionStatus(StrEnum):
    """Position status codes from 88-level values in POSREC."""

    ACTIVE = "A"
    CLOSED = "C"
    PENDING = "P"


class PositionRecord(BaseModel):
    """Full position record translated from COBOL POSITION-RECORD.

    Maps to POSREC.cpy copybook fields:
    - POS-KEY (composite key: portfolio_id + date + investment_id)
    - POS-DATA (position details)
    """

    # Key fields (POS-KEY)
    portfolio_id: str = Field(max_length=8, description="POS-PORTFOLIO-ID")
    date: str = Field(max_length=8, description="POS-DATE: YYYYMMDD")
    investment_id: str = Field(max_length=10, description="POS-INVESTMENT-ID")

    # Data fields (POS-DATA)
    quantity: Decimal = Field(max_digits=15, decimal_places=4, description="POS-QUANTITY S9(11)V9(4)")
    cost_basis: Decimal = Field(max_digits=15, decimal_places=2, description="POS-COST-BASIS S9(13)V9(2)")
    market_value: Decimal = Field(max_digits=15, decimal_places=2, description="POS-MARKET-VALUE S9(13)V9(2)")
    currency: str = Field(max_length=3, description="POS-CURRENCY")
    status: PositionStatus = Field(default=PositionStatus.ACTIVE, description="POS-STATUS: A/C/P")

    @property
    def composite_key(self) -> str:
        """Build the composite key matching VSAM KSDS key structure (18 bytes)."""
        return f"{self.portfolio_id}{self.date}{self.investment_id}"

    @property
    def unrealized_gain_loss(self) -> Decimal:
        """Calculate unrealized gain/loss."""
        return self.market_value - self.cost_basis

    @property
    def gain_loss_percent(self) -> Decimal:
        """Calculate gain/loss percentage."""
        if self.cost_basis == 0:
            return Decimal("0")
        return ((self.market_value - self.cost_basis) / self.cost_basis) * 100

File: frontend/src/types/position.ts
/**
 * Domain types for the Position Record.
 *
 * These mirror the COBOL copybook `src/copybook/common/POSREC.cpy`
 * (01 POSITION-RECORD). Field names are translated from COBOL hyphenated names
 * to camelCase, but the underlying semantics, lengths, and allowed values are
 * preserved so the structure can later be mapped 1:1 to the backend INQPORT
 * program / POSFILE VSAM records.
 */

/** POS-STATUS (X(1)) — 88-level values from POSREC.cpy. */
export type PositionStatus = 'A' | 'C' | 'P';

export const POSITION_STATUS_LABELS: Record<PositionStatus, string> = {
  A: 'Active',
  C: 'Closed',
  P: 'Pending',
};

export const POSITION_STATUSES = Object.keys(
  POSITION_STATUS_LABELS,
) as PositionStatus[];

/**
 * A single portfolio position.
 *
 * Monetary and quantity fields are carried as plain decimal strings (e.g.
 * "1234567.89") rather than JS numbers. POS-COST-BASIS / POS-MARKET-VALUE are
 * `S9(13)V9(2) COMP-3` and POS-QUANTITY is `S9(11)V9(4) COMP-3` — both exceed
 * the safe-integer range of an IEEE-754 double once scaled, so keeping the raw
 * decimal string preserves COMP-3 packed-decimal precision end to end
 * (see utils/decimal.ts).
 */
export interface Position {
  /** POS-PORTFOLIO-ID PIC X(8) — part of POS-KEY. */
  portfolioId: string;
  /** POS-DATE PIC X(8) — YYYYMMDD, part of POS-KEY. */
  date: string;
  /** POS-INVESTMENT-ID PIC X(10) — part of POS-KEY (the "Fund ID"). */
  investmentId: string;
  /**
   * Human-readable fund name for display.
   *
   * NOTE: POSREC.cpy has no fund-name field — POS-INVESTMENT-ID is the only
   * fund identifier. The name is an enrichment that would come from an
   * investment/security master file in production; it is supplied by the
   * fixture here so the inquiry table can show the "Fund Name" column.
   */
  fundName: string;
  /** POS-QUANTITY — signed decimal string with up to 4 fraction digits. */
  quantity: string;
  /** POS-COST-BASIS — signed decimal string with 2 fraction digits. */
  costBasis: string;
  /** POS-MARKET-VALUE — signed decimal string with 2 fraction digits. */
  marketValue: string;
  /** POS-CURRENCY PIC X(3) — ISO currency code. */
  currency: string;
  /** POS-STATUS. */
  status: PositionStatus;
  /** POS-LAST-MAINT-DATE PIC X(26). */
  lastMaintDate: string;
  /** POS-LAST-MAINT-USER PIC X(8). */
  lastMaintUser: string;
}

/** Search/filter parameters for the position inquiry. */
export interface PositionQuery {
  /** Filter by POS-STATUS; empty string means "all statuses". */
  status?: PositionStatus | '';
}
